torch/csrc/distributed/rpc/rref.cpp

```diff
@@ -1,119 +1,188 @@
-    // creator user, notify owner.
-    auto& agent = RRefContext::getInstance()->agent();
-    agent->send(
-        agent->getWorkerId(ownerId_),
-        ScriptRRefCreate(RRefForkData(ownerId_, rrefId_, forkId_).toIValue())
-            .toMessage());
-  } else {
-    AT_ERROR("Does not support sharing RRefs between users yet");
-  }
+  // Do nothing,
+  // (1) If this UserRRef is a fork of an existing RRef, RRefContext will send
+  //     a RREF_FORK_REQUEST message to the owner.
+  // (2) If this the creator UserRRef, ScriptRemoteCall or PythonRemoteCall will
+  //     properly notify the owner.
 }
 
-UserRRef::~UserRRef() {
+template <typename T>
+UserRRef<T>::~UserRRef() {
+  // TODO: queue this in RRefContext instead of doing it here.
   auto& ctx = RRefContext::getInstance();
   if (ctx->getWorkerId() != ownerId_) {
-    ctx->agent()->send(
-        ctx->agent()->getWorkerId(ownerId_),
-        ScriptRRefDelete(RRefForkData(ownerId_, rrefId_, forkId_).toIValue())
-            .toMessage());
+    auto fm = ctx->agent()->send(
+        ctx->agent()->getWorkerInfo(ownerId_),
+        RRefUserDelete(rrefId_, forkId_).toMessage());
+
+    fm->addCallback(
+        [](const Message& message) { RRefContext::handleException(message); });
   }
 }
 
-const ForkId& UserRRef::forkId() const {
+template <typename T>
+const ForkId& UserRRef<T>::forkId() const {
   return forkId_;
 }
 
-bool UserRRef::isOwner() const {
-  return false;
+template <>
+IValue UserRRef<IValue>::toHere() {
+  auto& agent = RRefContext::getInstance()->agent();
+  std::shared_ptr<FutureMessage> fm = agent->send(
+      agent->getWorkerInfo(ownerId_),
+      ScriptRRefFetchCall(rrefId()).toMessage());
+  const Message& message = fm->wait();
+  RRefContext::handleException(message);
+  auto srv = RRefFetchRet::fromMessage(message);
+  return srv.value();
 }
 
-IValue UserRRef::toHere() {
+template <>
+py::object UserRRef<py::object>::toHere() {
   auto& agent = RRefContext::getInstance()->agent();
   std::shared_ptr<FutureMessage> fm = agent->send(
-      agent->getWorkerId(ownerId_),
-      ScriptRRefFetchCall(id().toIValue()).toMessage());
-  auto srv = ScriptRRefFetchRet::fromMessage(fm->wait());
-  return srv.value();
+      agent->getWorkerInfo(ownerId_),
+      PythonRRefFetchCall(rrefId()).toMessage());
+  const Message& message = fm->wait();
+  RRefContext::handleException(message);
+  auto srv = RRefFetchRet::fromMessage(message);
+  return PythonRpcHandler::getInstance().deserialize(srv.value().toStringRef());
+}
+
+template class UserRRef<IValue>;
+template class UserRRef<py::object>;
+
+//////////////////////////  OwnerRRef  /////////////////////////////////////
+
+template <typename T>
+const T& OwnerRRef<T>::getValue() const {
+  // TODO: use callback to make this non-blocking
+  std::unique_lock<std::mutex> lock(mutex_);
+  valueCV_.wait(lock, [this] { return value_.has_value(); });
+  return value_.value();
```

Comment:
Is the function that produces the value guaranteed to run if you're blocking this thread?

Reply:
It should be, if there is no failures. Failures will be addressed in followup PRs.